Prevent host access on VIP addresses in proxy-mode=ipvs

uablrek opened the pull request:
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

In proxy-mode=ipvs ports on the host, e.g. ssh:22, can  be accessed from an external machine using an external VIP address or a service address.

#### Which issue(s) this PR fixes:

Fixes #72236

#### Special notes for your reviewer:

An attempt at solving this according to https://github.com/kubernetes/kubernetes/issues/72236#issuecomment-1057031916

* Create and maintain a KUBE-IPVS0-IPS ipset (hash:ip) that holds all addresses assigned to kube-ipvs0
* Add iptables rules as below
* Test really hard every case where both kubelet and kube-proxy updates the INPUT filter chain
```
iptables -A INPUT -m set --match-set KUBE-LOAD-BALANCER dst,dst -j ACCEPT
iptables -A INPUT -m set --match-set KUBE-CLUSTER-IP dst,dst -j ACCEPT
iptables -A INPUT -m set --match-set KUBE-IPVS0-IPS dst -j REJECT
```

I am very uncertain about the division of iptables handling between `kubelet` and `kube-proxy`. The filter-INPUT chain is from what I can see only updated by `kubelet`. Please advice on how to update it from `kube-proxy` without messing up.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

uablrek commented:
@andrewsykim Please review and advise about the update of the filter-INPUT chain.

The update to create a `KUBE-IPVS0-IPS` ipset was _much_ smaller than I feared :smile:

uablrek commented:
Unfortunately there is more to it. When the iptables rules are applied access to the host ports is rejected ok, but access to the service from main netns on the node doesn't work :disappointed: 

This is because traffic from main netns gets the VIP address as source. Example for 10.0.0.0 as VIP;
```
# ip ro show table local
local 10.0.0.0 dev kube-ipvs0 proto kernel scope host src 10.0.0.0 
...
```
This causes reply traffic with dest `10.0.0.0:ephemeral-port` to be rejected.

uablrek commented:
Adding a ctstate solves the problem;
```
# iptables -S INPUT 
-P INPUT ACCEPT
-A INPUT -m comment --comment "kubernetes health check rules" -j KUBE-NODE-PORT
-A INPUT -j KUBE-FIREWALL
-A INPUT -m set --match-set KUBE-LOAD-BALANCER dst,dst -j ACCEPT
-A INPUT -m set --match-set KUBE-CLUSTER-IP dst,dst -j ACCEPT
-A INPUT -m conntrack --ctstate NEW -m set --match-set KUBE-IPVS0-IPS dst -j REJECT --reject-with icmp-port-unreachable
```
But are all cases covered? How about UDP?

uablrek commented:
/test all

uablrek commented:
/test pull-kubernetes-verify

uablrek commented:
/test pull-kubernetes-e2e-gci-gce-ipvs

uablrek commented:
/test pull-kubernetes-e2e-gci-gce-ipvs

uablrek commented:
/assign @andrewsykim

aojea commented:
/test pull-kubernetes-e2e-gci-gce-ipvs


xref https://github.com/kubernetes/kubernetes/pull/108496#issuecomment-1059416264

Commit:
Prevent host access on VIP addresses in proxy-mode=ipvs

uablrek commented:
/retest

danwinship commented:
> ```
> iptables -A INPUT -m set --match-set KUBE-LOAD-BALANCER dst,dst -j ACCEPT
> iptables -A INPUT -m set --match-set KUBE-CLUSTER-IP dst,dst -j ACCEPT
> iptables -A INPUT -m set --match-set KUBE-IPVS0-IPS dst -j REJECT
> ```

In general, rather than adding arbitrary rules to the built-in chains, it is better to add a single rule like `-A INPUT -j KUBE-IPVS-FILTER` (via the `iptablesJumpChain` array) and then add rules to that to do what you want. This makes it easier to clean things up (via `iptablesCleanupChains`) because you can just flush the whole chain and any rules that jump to it, rather than needing to figure out exactly which rules to delete from `INPUT`.

> I am very uncertain about the division of iptables handling between `kubelet` and `kube-proxy`. The filter-INPUT chain is from what I can see only updated by `kubelet`. Please advice on how to update it from `kube-proxy` without messing up.

The ipvs proxy mostly doesn't use `filter` but the iptables proxy does. There is no problem with kubelet interactions, as long as you aren't trying to use the same chains.

> But are all cases covered? How about UDP?

Yes, conntrack keeps track of UDP packets and assumes that patterns that look like "connections" should be treated as connections.

danwinship commented on pkg/proxy/ipvs/ipset.go:
```diff
@@ -78,6 +78,9 @@ const (
 	kubeHealthCheckNodePortSet        = "KUBE-HEALTH-CHECK-NODE-PORT"
+
+	kubeIpvs0SetComment = "Addresses on interface kube-ipvs0"
+	kubeIpvs0Set        = "KUBE-IPVS0-IPS"
```
FWIW, `"kube-ipvs0"` is currently defined as a constant (`defaultDummyDevice`) in `proxier.go`, and the name kind of implies that the dummy device _could_ be something "non-default" instead, although in fact, there's no support for actually doing that. So anyway, if you're going to start explicitly referencing the interface name elsewhere, maybe `defaultDummyDevice` should be renamed to something that makes it clearer that it's not just a "default"? (in a separate commit)

danwinship commented on pkg/util/ipset/ipset.go:
```diff
@@ -169,6 +169,11 @@ func (e *Entry) Validate(set *IPSet) bool {
 	switch e.SetType {
+	case HashIP:
+		//check if IP and Protocol of Entry is valid.
+		if valid := e.checkIPandProtocol(set); !valid {
```
`hash:ip` has no protocol

danwinship replied:
(this comment still applies... it looks like `checkIPandProtocol()` will accept an `Entry` with no `Protocol`, but it will fill in a default value of `"TCP"`, which isn't really what you want... there ought to be a `checkIP()` function that doesn't do that...)

danwinship commented on pkg/util/ipset/ipset.go:
```diff
@@ -219,6 +224,10 @@ func (e *Entry) Validate(set *IPSet) bool {
+	case HashIP:
+		// Entry{192.168.1.1, udp} -> 192.168.1.1,udp
+		// Entry{192.168.1.2, tcp} -> 192.168.1.2,tcp
+		return fmt.Sprintf("%s,%s", e.IP, e.Protocol)
```
likewise

uablrek commented:
Updated after review. Iptables now looks like;
```
# iptables -L -nv
Chain INPUT (policy ACCEPT 85 packets, 24249 bytes)
 pkts bytes target     prot opt in     out     source               destination         
 2509   51M KUBE-IPVS-FILTER  all  --  *      *       0.0.0.0/0            0.0.0.0/0            /* kubernetes ipvs access filter */
 2508   51M KUBE-NODE-PORT  all  --  *      *       0.0.0.0/0            0.0.0.0/0            /* kubernetes health check rules */
 2707   51M KUBE-FIREWALL  all  --  *      *       0.0.0.0/0            0.0.0.0/0           
...
Chain KUBE-IPVS-FILTER (1 references)
 pkts bytes target     prot opt in     out     source               destination         
    0     0 ACCEPT     all  --  *      *       0.0.0.0/0            0.0.0.0/0            match-set KUBE-LOAD-BALANCER dst,dst
    0     0 ACCEPT     all  --  *      *       0.0.0.0/0            0.0.0.0/0            match-set KUBE-CLUSTER-IP dst,dst
    0     0 ACCEPT     all  --  *      *       0.0.0.0/0            0.0.0.0/0            match-set KUBE-EXTERNAL-IP dst,dst
    0     0 REJECT     all  --  *      *       0.0.0.0/0            0.0.0.0/0            ctstate NEW match-set KUBE-IPVS-IPS dst reject-with icmp-port-unreachable
...
```

I have added the KUBE-IPVS-FILTER chain in the `iptablesCleanupChains` array.

uablrek commented:
I tested to stop `kube-proxy` and then run;
```
# kube-proxy --cleanup
I0622 11:52:07.060172   25154 server.go:231] "Warning, all flags other than --config, --write-config-to, and --cleanup are deprecated, please begin using a config file ASAP"
```
The `KUBE-IPVS-FILTER` and it's reference was correctly removed from `iptables` and `ip6tables`.

uablrek commented:
In my test env I test from an external machine with;
```
for adr in 1000::1 10.0.0.1; do
  echo "" | nc -w1 $adr 22 > /dev/null && tdie "Access works to [$adr]"
done
```
1000::1 and 10.0.0.1 are loasBalancerIPs. Fails on v1.24.2 and succeeds with this PR.

uablrek commented:
/test pull-kubernetes-e2e-kind-ipv6

danwinship commented on pkg/proxy/ipvs/ipset.go:
```diff
@@ -78,6 +78,9 @@ const (
 	kubeHealthCheckNodePortSetComment = "Kubernetes health check node port"
 	kubeHealthCheckNodePortSet        = "KUBE-HEALTH-CHECK-NODE-PORT"
+
+	kubeIpvsSetComment = "Addresses on the ipvs interface"
```
"`IPVS`" not "`Ipvs`", here and in all the other const/variable names. (Acronyms in golang/kubernetes names never get titled-cased; they're always either all-uppercase or all-lowercase.)

danwinship reviewed (approved):
lgtm other than nitpicks.

I'm not that experienced with the ipvs code so it would be nice if someone else reviewed it too, but it seems like ipvs reviewers are in short supply right now so that may not be possible...

danwinship commented:
ah, it seems that only Tim can approve `pkg/util/ipset` anyway... which is odd...
/assign @thockin

uablrek commented:
Fixed the last comments. If this PR is merged it should be back-ported IMHO. This is a security problem that has been unresolved for too long.

danwinship commented:
/lgtm
/approve

danwinship commented:
The "drop packets to ipvs0 addresses" rule wouldn't be useful in FORWARD, but it would be harmless...

uablrek commented:
@danwinship Should I remove KUBE-IPVS-FILTER and use KUBE-PROXY-FIREWALL?

It would be cleaner, but almost everything passes FORWARD, so...?

thockin commented on pkg/proxy/ipvs/proxier.go:
```diff
@@ -1779,6 +1790,22 @@ func (proxier *Proxier) writeIptablesRules() {
 
+	// Add rules to the filter/KUBE-IPVS-FILTER chain to prevent access to ports on the host through VIP addresses.
+	// https://github.com/kubernetes/kubernetes/issues/72236
+	proxier.filterRules.Write(
```
@danwinship We have historically NOT used ipset in iptables mode, but maybe that's something we should consider?

thockin reviewed (commented):
Thanks!

/lgtm
/approve
